Changes to the Log to support log segmentation

vgkholla opened the pull request:
This commit introduces the changes required for the log to use `LogSegment` instances. It allows the log to be segmented into multiple segments. However, the index, hard delete and `BlobStore` do not yet completely support segmentation and hence the log cannot yet be segmented. The current configs make sure that segments aren't created

`./gradlew build && ./gradlew test` succeeded.
Code formatted.

**Primary reviewers: @pnarayanan, @nsivabalan**
**Expected time to review: PROD: 60-90 mins TEST: 30-60 mins**

**Unit testing**

| Class | Class, % | Method, % | Line, % |
| --- | --- | --- | --- |
| `LogSegment` | 100% (1/ 1) | 100% (18/ 18) | 100% (93/ 93) |
| `Log` | 100% (1/ 1) | 100% (21/ 21) | 99.1% (110/ 111) |
| `LogSegmentNameHelper` | 100% (3/ 3) | 93.8% (15/ 16) | 95.7% (45/ 47) |
| `Offset` | 100% (1/ 1) | 88.9% (8/ 9) | 97.1% (34/ 35) |

`LogSegmentNameHelper` is a static class so the "method" not covered is its constructor.
Method and line not covered in `Offset` is `toString()`. 
All other classes with new logic have all lines including all if-else branches covered.

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,304 @@
+   * @throws IllegalStateException if there no more capacity in the log.
+   * @throws IOException if any I/O error occurred as a part of ensuring capacity.
+   */
+  private void ensureCapacity(long writeSize)
```
this method doesn't just ensure capacity alone. It also allocates next segment as well. Can we reflect that in method name. Ideally I would prefer to move the logic of allocating new segment into rollover method if you want to keep this method just to ensure capacity

vgkholla replied:
Allocating the next segment is the act of ensuring capacity.

pnarayanan replied:
+1. could you add more to the name and javadoc?

vgkholla replied:
Added to java doc. What do you guys think the name should be ?

vgkholla replied:
@nsivabalan does not always allocate a segment - in fact does not allocate one most of the times.

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/LogSegmentNameHelper.java:
```diff
@@ -0,0 +1,162 @@
+   * @return the relative position of the log segment.
+   */
+  static long getPosition(String name) {
+    if (name.isEmpty()) {
```
why do we need to check for empty? If we expect programmatic errors, then we should also be cautious for name being null

vgkholla replied:
I don't think there is any difference b/w null and empty here in the sense that both throw exceptions (which is what is required).

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/LogSegmentNameHelper.java:
```diff
@@ -0,0 +1,162 @@
+   * @return the generation number of the log segment.
+   */
+  static long getGeneration(String name) {
+    if (name.isEmpty()) {
```
same as above

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/LogSegment.java:
```diff
@@ -29,25 +35,34 @@
    * @throws IOException if the file cannot be read or created
    */
-  LogSegment(String name, File file, long capacityInBytes, StoreMetrics metrics)
+  LogSegment(String name, File file, long capacityInBytes, StoreMetrics metrics, boolean writeHeader)
```
I feel that we should also append totalCapacity of the LogSegment while serializing the file.

vgkholla replied:
Don't think it is required.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/BlobStore.java:
```diff
@@ -178,7 +175,8 @@ public void put(MessageWriteSet messageSetToWrite)
         }
-        long writeStartOffset = log.getLogEndOffset();
+        // TODO (Index changes): Working under the assumption that there is only one log segment.
+        long writeStartOffset = log.getLogEndOffset().getOffset();
```
Done.

vgkholla commented on ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageFormatWriteSet.java:
```diff
@@ -55,8 +50,12 @@ public MessageFormatWriteSet(InputStream streamToWrite, List<MessageInfo> stream
     ReadableByteChannel readableByteChannel = Channels.newChannel(streamToWrite);
-    writeChannel.appendFrom(readableByteChannel, sizeToWrite);
-    return sizeToWrite;
+    long sizeWritten = 0;
```
Each blob will be written in isolation with this change.

vgkholla replied:
@pnarayanan Yes that remains the same because index entries are not written until all the whole write has succeeded.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegmentNameHelper.java:
```diff
@@ -0,0 +1,162 @@
+ * The file name is a combination of the segment name and a suffix "_log"
+ */
+class LogSegmentNameHelper {
+  static final String SUFFIX = BlobStore.SEPARATOR + "log";
```
Since it used in both log and index, I think this is appropriate.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/LogSegmentNameHelper.java:
```diff
@@ -0,0 +1,162 @@
+    if (name.isEmpty()) {
+      throw new IllegalArgumentException("Name provided cannot be empty");
+    }
+    return Long.parseLong(name.substring(name.indexOf(BlobStore.SEPARATOR) + 1, name.length()));
```
removed.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,304 @@
+   * Sets the active segment in the log.
+   * @param name the name of the log segment that is to be marked active.
+   */
+  void setActiveSegment(String name) {
```
yup - the index has to use this.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,304 @@
+   */
+  long getUsedCapacity() {
+    long usedCapacity = 0;
+    for (Map.Entry<String, LogSegment> entry : segmentsByName.entrySet()) {
```
Yes - it is hard to keep this updated for compaction. Also the number of segments is finite and this loop will not take very long.

vgkholla replied:
I don't think this needs to be done right away. Once we have compaction in place, if we can do this intelligently without making anything go awry, we can look at it then. For now, I think the optimization is not required.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,304 @@
+    long usedCapacity = 0;
+    for (Map.Entry<String, LogSegment> entry : segmentsByName.entrySet()) {
+      LogSegment segment = entry.getValue();
+      if (segment != activeSegment) {
```
Yes - want to compare the actual references here.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,304 @@
-      metrics.overflowWriteError.inc(1);
-      throw new IllegalArgumentException("Log : " + file.getAbsolutePath() + " error trying to append to log " +
-          "from channel since new data size " + size + "exceeds total log size " + capacityInBytes);
+    for (Map.Entry<String, LogSegment> entry : segmentsByName.entrySet()) {
```
the index calls it when persisting index segments.

vgkholla replied:
@pnarayanan the index calls this function.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,304 @@
-      throw new IllegalArgumentException("Log : " + file.getAbsolutePath() + " error trying to write to log " +
-          "from channel since new data size " + size + "exceeds log end offset " + currentWriteOffset.get());
+    for (Map.Entry<String, LogSegment> entry : segmentsByName.entrySet()) {
+      entry.getValue().close();
```
I think its fine to throw as soon as one of them has a problem.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,304 @@
+  private void rollOverIfRequired(long writeSize)
       throws IOException {
-    fileChannel.force(true);
+    if (activeSegment.getCapacityInBytes() - activeSegment.getEndOffset() < writeSize) {
```
Currently, I don't think it is required (also that is derived data).

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,304 @@
-      throw new IllegalArgumentException("Log : " + file.getAbsolutePath() + " error trying to read outside " +
-          "the log range. log end position " + sizeInBytes() + " input buffer size " + buffer.remaining());
+    if (segmentsByName.higherEntry(activeSegment.getName()) != null) {
+      return;
```
This will happen if a crash occurs between a segment being allocated and the operation that caused the allocation being started.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,304 @@
+          // no headers
+          segment = new LogSegment(name, segmentFile, segmentCapacity, metrics, false);
+        } else {
+          segment = new LogSegment(name, segmentFile, metrics);
```
The config is useful only on first startup. When subsequent segments are created, their capacity is determined using the capacity of the active segment and not using the config.

vgkholla replied:
@pnarayanan As discussed offline, there is no situation where the log segments are not the same size since the code that allocates the log segments is in the same class - so it would be a programming bug if it ended up otherwise (and there is no "fix"). Moreover, I think it is an unnecessary/misleading check that might make it seem like it could ever happen. It would be akin to checking that index and log match for example - something we do not do.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,304 @@
+    if (activeSegment.getCapacityInBytes() - activeSegment.getEndOffset() < writeSize) {
+      ensureCapacity(writeSize);
+      // this cannot be null since capacity has either been ensured or has thrown.
+      LogSegment nextActiveSegment = segmentsByName.higherEntry(activeSegment.getName()).getValue();
```
so this is not necessarily `segmentsByName.lastEntry.getValue()`?

vgkholla replied:
DOne.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,304 @@
+    this.capacityInBytes = totalCapacityInBytes;
     this.metrics = metrics;
+    remainingUnallocatedCapacity = new AtomicLong(totalCapacityInBytes);
+    long numSegments = (totalCapacityInBytes - 1) / segmentCapacityInBytes + 1;
```
since you ensure `totalCapacityInBytes` is a multiple of `segmentCapacityInBytes`, why not keep this simple and write `numSegments = totalCapacityInBytes/segmentCapacityInBytes`?

vgkholla replied:
Done.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,304 @@
+    // all segments should be the same size.
+    if (totalCapacityInBytes % segmentCapacity != 0) {
+      throw new IllegalArgumentException(
+          "Capacity of log [" + totalCapacityInBytes + "] should be a perfect multiple of segment capacity ["
```
"perfect multiple" == "multiple"
(just saying)

vgkholla commented:
@nsivabalan comments addressed

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,304 @@
+    }
+    // all segments are (should be) the same size.
+    long segmentCapacity = activeSegment.getCapacityInBytes();
+    if (segmentCapacity > remainingUnallocatedCapacity.get()) {
```
this is not very intuitive. far more intuitive to check for `remainingUnallocatedCapacity.get()  == 0`. Isn't that really what you are trying to do here? Might be even simpler to just use `remainingUnallocatedSegments` as the value only gets decremented in multiples of `segmentCapacity`.

vgkholla replied:
Changed.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/LogSegmentNameHelper.java:
```diff
@@ -0,0 +1,162 @@
+    }
+    return filename.substring(0, filename.length() - SUFFIX.length());
+  }
+}
```
do you see a use for a validate() method? - that validates that a segment name adheres to the convention followed?
Also I feel that we should have a versioning somewhere. For version 0, you overcome the difficulty by assuming that existing one segment logs are actually old style segments. what if tomorrow one feels the need to add another field to the name?

vgkholla replied:
@nsivabalan asked for a validation method too. I have added one when the filename is converted to the name - I think that is a good point to add the validation. In the other functions, I think it would be excessive.

I thought about the versioning and I feel like the version inside the log segment can act as versioning for the log segment name also (similar to the index segment whose name does not have a versioning). If required later on, the log segment can have a `readHeader()` method that the log (or anyone else) can use to determine versioning and other things. I don't think the name itself needs the versioning.

pnarayanan replied:
version in the header is better as long as we recognize that log name format is part of the version and if that is changed, the version needs to be changed as well.

vgkholla replied:
Added a comment in `LogSegmentNameHelper`.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/BlobStore.java:
```diff
@@ -334,4 +333,17 @@ private void checkStarted()
+   * Sets the end offsets and states for all the segments (by setting the active segment).
+   * @throws IOException if there is an I/O error while setting the end offsets.
+   */
+  private void setSegmentStatesAndEndOffsets()
```
Yes that's correct - the presence of this function is (also was) extraneous because the index already does this. Depending how the index evolves, I might get rid of this function.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,304 @@
+          "One of totalCapacityInBytes [" + totalCapacityInBytes + "] or " + "segmentCapacityInBytes ["
+              + segmentCapacityInBytes + "] is <=0");
+    }
+    long segmentCapacity = Math.min(totalCapacityInBytes, segmentCapacityInBytes);
```
Not always. In case someone is not interested in segmentation, they would set the config to `Long.MAX_VALUE`.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,304 @@
+  /**
+   * Sets the active segment in the log.
+   * @param name the name of the log segment that is to be marked active.
+   */
```
The index will call this function to set the current active segment. That situation will arise when a segment has been allocated for an operation but a crash occurs just before the operation can actually proceed. As discussed, I will work on maintaining the invariant that the active segment is always the last segment.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,304 @@
+   * {@link LogSegment} instances that are not fully filled.
+   */
+  long getUsedCapacity() {
+    long usedCapacity = 0;
```
Will change.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,304 @@
+   * @return the number of valid segments starting from the first segment.
+   */
+  long getSegmentCount() {
+    return segmentsByName.size();
```
I'm keeping this open to see how this evolves with compaction (same for the index). If something simpler can be done, we can optimize then.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,304 @@
+      throws IOException {
+    File dir = new File(dataDir);
+    File[] segmentFiles = dir.listFiles(LogSegmentNameHelper.LOG_FILE_FILTER);
+    if (segmentFiles == null || segmentFiles.length == 0) {
```
Done.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,304 @@
     }
-    logger.trace("Log : {} bytes written to the log from read channel at {}, bytesWritten: {}", file.getAbsolutePath(),
-        offset, bytesWritten);
+    activeSegment = segmentsByName.firstEntry().getValue();
```
Changed to last entry. The active segment is set by the index.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,304 @@
 
-  public void flush()
+  /**
+   * Rolls the log segment over if required. Also ensures enough capacity.
```
Changed.
Ensures capacity by calling `ensureCapacity()`. Is that what your question is about?

pnarayanan replied:
ensures there is enough capacity in the log? in the segment?

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,304 @@
+   * Rolls the log segment over if required. Also ensures enough capacity.
+   * @param writeSize the size of the incoming write.
+   * @throws IllegalArgumentException if the {@code writeSize} is greater than a single segment's size
+   * @throws IllegalStateException if there no more capacity in the log.
```
Fixed.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,304 @@
+   * @param writeSize the size of the incoming write.
+   * @throws IllegalArgumentException if the {@code writeSize} is greater than a single segment's size
+   * @throws IllegalStateException if there no more capacity in the log.
+   * @throws IOException if any I/O error occurred as a part of ensuring capacity.
```
Fixed.

vgkholla commented:
@pnarayanan @nsivabalan Addressed comments

1. Made the load segments v/s create new segments path clearer
2. Made sure the active segment is the last segment at any point in time
3. Added some validation to segment file names

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,304 @@
-    fileChannel.force(true);
+    if (activeSegment.getCapacityInBytes() - activeSegment.getEndOffset() < writeSize) {
+      ensureCapacity(writeSize);
+      // this cannot be null since capacity has either been ensured or has thrown.
```
Decided not to do this after discussion with @pnarayanan because this is in the critical operation path.

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/LogSegmentNameHelper.java:
```diff
@@ -0,0 +1,174 @@
+ * <p/>
+ * The file name is a combination of the segment name and a suffix "_log"
+ */
+class LogSegmentNameHelper {
```
this is actually a helper class for version 1 of LogSegment. May be good to reflect that in class name or elsewhere.

vgkholla replied:
This is actually meant to take versions - it hasn't been added because it is not required here but the class can be extended to handle versioning - so i don't want to limit its scope.

vgkholla replied:
Added a comment that addresses this

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,392 @@
+    if (!segmentsByName.containsKey(name)) {
+      throw new IllegalArgumentException("There is no log segment with name: " + name);
+    }
+    Map.Entry<String, LogSegment> higherEntry = segmentsByName.higherEntry(name);
```
I think this is fine. Don't think one is better than the other.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,392 @@
+      throw new IllegalArgumentException("One of the log segments provided [" + o1.getName() + ", " + o2.getName() +
+          "] does not belong to this log");
+    }
+    LogSegment firstSegment = segmentsByName.get(o1.getName());
```
Done.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,392 @@
+          "] does not belong to this log");
+    }
+    LogSegment firstSegment = segmentsByName.get(o1.getName());
+    LogSegment secondSegment = segmentsByName.get(o2.getName());
```
They are not start and end offsets because start and end gives the impression that start <= end (which is not the case here). I have avoided naming them start-end for that reason.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,392 @@
+    Offset lower = higher == o1 ? o2 : o1;
+    LogSegment segment = segmentsByName.higherEntry(lower.getName()).getValue();
+    // segment capacity is the same for every segment.
+    long difference = segment.getCapacityInBytes() - lower.getOffset();
```
all segments are of the same capacity (comment above the line was put to avoid the confusion).

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,392 @@
+    for (File segmentFile : segmentFiles) {
+      String name = LogSegmentNameHelper.nameFromFilename(segmentFile.getName());
+      LogSegment segment;
+      if (name.isEmpty()) {
```
I usually prefer to not repeating code. I think the current code is also simple - let me know if you feel strongly otherwise.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,392 @@
+      logger.info("Rolling over writes to {} from {} on write of data of size {}. End offset was {} and capacity is {}",
+          nextActiveSegment.getName(), activeSegment.getName(), writeSize, activeSegment.getEndOffset(),
+          activeSegment.getCapacityInBytes());
+      activeSegment = nextActiveSegment;
```
Not sure why?

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Offset.java:
```diff
@@ -36,11 +36,11 @@
-   * @throws IllegalArgumentException if {@code name} is {@code null} or empty or {@code offset} < 0.
+   * @throws IllegalArgumentException if {@code name} is {@code null} or {@code offset} < 0.
    */
   Offset(String name, long offset) {
```
I understand the intent but I don't think `offset.getOffsetInto()` or `offset.getLogOffset()` sounds any better. @pnarayanan pls comment.

vgkholla commented:
3d82a6a is the latest commit - the rest is a rebase after some merge conflicts were resolved.

nsivabalan commented on ambry-store/src/test/java/com.github.ambry.store/LogTest.java:
```diff
@@ -14,214 +14,645 @@
+      log.setActiveSegment(activeSegmentName);
+      // all segment files from segmentIdxToMarkActive + 1 to expectedSegmentNames.size() - 1 will be freed.
+      List<String> prunedSegmentNames = expectedSegmentNames.subList(0, segmentIdxToMarkActive + 1);
+      checkLog(log, Math.min(logCapacity, segmentCapacity), numSegments, prunedSegmentNames);
```
will sync up with you offline to understand this better.

vgkholla commented on ambry-store/src/test/java/com.github.ambry.store/LogSegmentNameHelperTest.java:
```diff
@@ -0,0 +1,292 @@
+  @Test
+  public void filenameFilterTest()
+      throws IOException {
+    File tempFile = File.createTempFile("ambry", ".tmp");
```
Removed.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,392 @@
+      String name = LogSegmentNameHelper.nameFromFilename(segmentFile.getName());
+      LogSegment segment;
+      if (name.isEmpty()) {
+        totalSegments = 1;
```
It cannot be created - I don't want to give the impression that it can be. I don't think it is possible to check for everything i.e. why would there be other files with "log_current" if there is no code creating it? Would prefer not to add those checks. Let me know if you feel strongly about this.

vgkholla commented on ambry-store/src/test/java/com.github.ambry.store/LogSegmentTest.java:
```diff
@@ -156,22 +159,23 @@ public void viewAndRefCountTest()
-      // cannot set end offset to illegal values (< 0 or > file size)
-      int[] invalidOffsets = {-1, (int) (segment.sizeInBytes() + 1)};
+      // cannot set end offset to illegal values (< initial offset or > file size)
+      int[] invalidOffsets = {(int) (writeStartOffset - 1), (int) (segment.sizeInBytes() + 1)};
```
-1 == writeStartOffset - 1

Why 1?

vgkholla commented on ambry-store/src/test/java/com.github.ambry.store/LogSegmentTest.java:
```diff
@@ -350,28 +363,103 @@ public void writeFromTest()
+    }
+
+    // unknown version
+    LogSegment segment = getSegment("log_current", STANDARD_SEGMENT_SIZE, true);
```
sure

vgkholla commented on ambry-store/src/test/java/com.github.ambry.store/LogSegmentTest.java:
```diff
@@ -350,28 +363,103 @@ public void writeFromTest()
+    writeHeader(segment, header);
+    try {
+      new LogSegment(name, file, metrics);
+      fail("Construction should have failed because crc check should have failed");
```
Not sure what the point of corrupting the CRC is? The test here is to check that we do a CRC match - this means that the data has to be corrupted, not the CRC.

vgkholla commented:
Rebased once again to get rid of merge conflicts. d784c29 is the latest commit

vgkholla commented:
@nsivabalan addressed comments.

pnarayanan commented on ambry-metrics/src/main/java/com.github.ambry.metrics/MetricsRegistryMap.java:
```diff
@@ -1,115 +0,0 @@
- * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
- * You may obtain a copy of the License at
- *
```
were all these unused classes?

vgkholla replied:
This was a part of the `StoreManager` PR whose commit was removed - Casey has just put up a PR reapplying that commit - should go away once that is merged

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,380 @@
+          "There is no more capacity left in [" + dataDir + "]. Max capacity is [" + capacityInBytes + "]");
+    }
+    // all segments are (should be) the same size.
+    long segmentCapacity = activeSegment.getCapacityInBytes();
```
Should `getCapacity()` return `actual_capacity - HEADER_SIZE`? What feels cleaner?

vgkholla replied:
the actual_capacity is cleaner I think. Doing a `- HEADER_SIZE` is a specialized use case that is used by components that are very aware of how the log segment works.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -1030,13 +1037,13 @@ public void write()
 
           // flush the log to ensure everything till the fileEndPointerBeforeFlush is flushed
-          log.flush();
+          logSegment.flush();
```
Would need to flush more than this segment in the general case, in case rollover happened twice in the flush interval. Wouldn't you?

vgkholla replied:
Yes - the changes in the index in this patch are done assuming one single segment. The next patch (that has index changes) will flush the `Log`.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,380 @@
+    }
+    Map.Entry<String, LogSegment> higherEntry = segmentsByName.higherEntry(name);
+    while (higherEntry != null) {
+      free(higherEntry.getValue());
```
Done. The file is currently deleted (impl of `free()` right now) but once the `DiskManager` implements pooling, the file will be returned to the pool.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,380 @@
-              buffer.remaining() + " exceeds total log size " + capacityInBytes);
+          "One of the log segments provided [" + o1.getName() + ", " + o2.getName() + "] does not belong to this log");
+    }
+    if (o1.getOffset() > firstSegment.getEndOffset() || o2.getOffset() > secondSegment.getEndOffset()) {
```
I have made it a capacity check here too in order to avoid any confusion.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,380 @@
+    LogSegment segment = segmentsByName.higherEntry(lower.getName()).getValue();
+    // segment capacity is the same for every segment.
+    long difference = segment.getCapacityInBytes() - lower.getOffset();
+    while (!segment.getName().equals(higher.getName())) {
```
I see - I did not consider lookup cost. Modified this now.

vgkholla replied:
Thanks for the info btw !

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,380 @@
-      throw new IllegalArgumentException("Log : " + file.getAbsolutePath() + " error trying to write to log " +
-          "from channel since new data size " + size + "exceeds log end offset " + currentWriteOffset.get());
+  void flush() throws IOException {
+    for (LogSegment segment : segmentsByName.values()) {
```
Added a TODO.

vgkholla commented:
@pnarayanan Addressed comments.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -13,161 +13,390 @@
+  }
+
+  /**
+   * Closes the Log and all its segments.
```
Does this guarantee a `flush` and `close`?

vgkholla replied:
I assumed that filechannel's close will provide a flush also but that is probably a bad assumption. Will flush explicitly.

vgkholla commented:
@pnarayanan 
Regarding your questions:
1. Do you have a test for the case where after recovery there is an extra log segment?
**Yes - those lines (and case) are covered.** 
2. In your bad tests, do you have a test where the Log is instantiated using a directory that has bad/invalid files?
**What kind of bad/invalid files? There are tests for when it is not a directory but if some file is named with _log or log_current, it is assumed to be valid.**
3. Is there a bad argument test where you try to write to the Log more than what a single log segment can hold?
**Yes. This is covered (`appendErrorCasesTest()`)**

nsivabalan commented:
Looks good to me.
